### PR TITLE
Mark Top Level Span

### PR DESCRIPTION
**Pull request by mabdinur**
## Description

Currently spans are marked as "top level" in the agent. When partial flushing is enabled this leads to errors in metric calculations/aggregations since some "chunk root" spans are incorrectly marked as "top level". The proposed change is outlined below.

This PR is a draft meant to receive feedback. Integration tests will be added after this approach is approved.

Top Level Span Calculation in the agent: https://github.com/DataDog/datadog-agent/blob/0454961e636342c9fbab9e561e6346ae804679a9/pkg/trace/traceutil/span.go#L16-L17

Rules to set the _dd.top_level Span metric to 0/1 in the Python library.

Always set _dd.top_level to 1 for local root span 
In each integration
If the integration has a service name different than it’s parent
Set _dd.top_level to 1  
On a partial flush

## Checklist
- [ ] Added to the correct milestone.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

File: ddtrace/internal/processor/trace.py [hunk 1]
```diff
@@ -159,6 +160,17 @@ def on_span_finish(self, span):
...
+                if not spans:
+                    return
+                chunk_root = spans[0]
+                if chunk_root is chunk_root._local_root:
+                    chunk_root.set_metric("_dd.top_level", 1)
+                elif chunk_root._parent and chunk_root._parent.service != chunk_root.service:
+                    chunk_root.set_metric("_dd.top_level", 1)
+                elif should_partial_flush:
+                    chunk_root.set_metric("_dd.top_level", 0)
+
```
Comment (mabdinur) [inline ddtrace/internal/processor/trace.py:173]: Should I move this logic to a trace processor? It seems out of place here

**Comment by mabdinur** [comment]
Quite a few integration tests are failing with the following error: `Span data mismatch.
❌ Key 'metrics._dd.top_level' not expected.` Which is expected. After this approach is approved I'll will update these tests

File: ddtrace/internal/processor/trace.py [hunk 2]
```diff
@@ -159,6 +159,16 @@ def on_span_finish(self, span):
...
 
+                # mark top level span
+                if spans:
+                    chunk_root = spans[0]
+                    if chunk_root is chunk_root._local_root:
+                        chunk_root.set_metric("_dd.top_level", 1)
+                    elif chunk_root._parent and chunk_root._parent.service != chunk_root.service:
+                        chunk_root.set_metric("_dd.top_level", 1)
+                    elif should_partial_flush:
+                        chunk_root.set_metric("_dd.top_level", 0)
```
Comment (mabdinur) [inline ddtrace/internal/processor/trace.py:170]: Should I move this logic to a trace processor (ex. TraceTagsProcessor)? It seems out of place here

Comment (brettlangdon) [reply ddtrace/internal/processor/trace.py:170]: or it could be it's own trace processor `TopLevelTraceProcessor` or something. then we'll get better separation of concerns.

what do you think?

Comment (brettlangdon) [reply ddtrace/internal/processor/trace.py:170]: besides that, I think this logic makes sense. Originally I was thinking we'd want this in `Tracer.start_span`, or even an `on_start_span` hook, but this approach is likely more performant since we only need to look at and consider the spans that the trace agent currently messes up when partial flushing is enabled.

**Review by brettlangdon — CHANGES_REQUESTED** [review]
big picture looks good, we could shuffle around where this code lives, but the logic seems correct to me at first glance.

main thing missing is test cases. if we could get some functional and integration tests together then 👍🏻

File: ddtrace/internal/processor/trace.py (hunk 2 above)
Comment (Kyle-Verhoog) [reply ddtrace/internal/processor/trace.py:170]: Yes a `TraceProcessor` would be preferred here. We'd lose the information about whether it's a partial flush or not but I think we can assume that if the first span in the chunk is not the local root then we are partial flushing anyway (there's no other case when this happens).

File: ddtrace/internal/processor/trace.py [hunk 3]
```diff
@@ -159,6 +159,16 @@ def on_span_finish(self, span):
                     except Exception:
                         log.error("error applying processor %r", tp, exc_info=True)
 
+                # mark top level span
+                if spans:
+                    chunk_root = spans[0]
+                    if chunk_root is chunk_root._local_root:
+                        chunk_root.set_metric("_dd.top_level", 1)
+                    elif chunk_root._parent and chunk_root._parent.service != chunk_root.service:
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/processor/trace.py:167]: we should merge this condition with the one above

**Review by Kyle-Verhoog — COMMENTED** [review]
apologies for the super late review 😭😭😭 

Just moving it to a `TraceProcessor` and adding some tests and we're good!

File: ddtrace/internal/processor/trace.py [hunk 4]
```diff
@@ -68,6 +68,25 @@ def process_trace(self, trace):
...
+class TraceTopLevelSpanProcessor(TraceProcessor):
+    """Processor marks spans as top level"""
+
+    def process_trace(self, trace):
+        # type: (List[Span]) -> Optional[List[Span]]
+
+        if trace:
+            chunk_root = trace[0]
+            if chunk_root is chunk_root._local_root:
+                chunk_root.set_metric("_dd.top_level", 1)
```
Comment (brettlangdon) [inline ddtrace/internal/processor/trace.py:81]: I wonder if we can skip this, since this is the scenario that is currently implied by the trace agent?

or better to set explicitly so there is less ambiguity?

**Review by brettlangdon — COMMENTED** [review]
one non-blocking comment, but otherwise looks like an appropriate implementation to me

File: ddtrace/internal/processor/trace.py (hunk 4 above)
Comment (mabdinur) [reply ddtrace/internal/processor/trace.py:81]: I think it's better to reduce ambiguity

**Review by brettlangdon — CHANGES_REQUESTED** [review]
It seems we also need to add this header to HTTP requests to the trace agent.

https://github.com/DataDog/datadog-agent/blob/ec8411214a68e77a60b9d3f7404ece8eb77f5956/pkg/trace/api/api.go#L352-L354

File: tests/snapshots/tests.contrib.grpc.test_grpc.test_method_service.snap [hunk 5]
```diff
@@ -18,9 +18,10 @@
            "grpc.method.service" "Servicer"
            "grpc.method.name" "Handler"}
    "metrics" {"_dd.agent_psr" 1.0
+              "_dd.top_level" 1
               "_dd.measured" 1
               "_sampling_priority_v1" 1
-              "system.pid" 18581
+      "system.pid" 18581
```
Comment (brettlangdon) [inline tests/snapshots/tests.contrib.grpc.test_grpc.test_method_service.snap:24]: unnecessary change?

File: tests/snapshots/tests.integration.test_integration_snapshots.test_tracer_trace_across_multiple_forks.snap [hunk 6]
```diff
@@ -36,5 +38,6 @@
            "duration" 86000
            "meta" {"runtime-id" "859c494affcb4933a1e15b6c4d2561b8"}
            "metrics" {"_sampling_priority_v1" 1
+                      "_dd.top_level" 1
```
Comment (brettlangdon) [inline tests/snapshots/tests.integration.test_integration_snapshots.test_tracer_trace_across_multiple_forks.snap:41]: This doesn't seem correct does it?

@Kyle-Verhoog what do you think the expected behavior is here?

Same with the previous "across fork" one as well.

**Review by brettlangdon — COMMENTED** [review]
I don't see any tests cases that `_dd.top_level = 0`

we need to add test cases for the new trace processor, and we probably want to update any partial flushing test cases to validate top level as they are the primary reason to need to mark as `0`

I am also curious about expected behavior from tracing across forks/processes, since it doesn't seem like those should be top level of 1.

**Comment by mabdinur** [comment]
> I don't see any tests cases that `_dd.top_level = 0`
> 
> we need to add test cases for the new trace processor, and we probably want to update any partial flushing test cases to validate top level as they are the primary reason to need to mark as `0`
> 
> I am also curious about expected behavior from tracing across forks/processes, since it doesn't seem like those should be top level of 1.

I removed the case that sets top level to 0 since it was unnecessary

**Comment by brettlangdon** [comment]
> I removed the case that sets top level to 0 since it was unnecessary

we need to have _something_ to validate this code is working as intended. Updating the fixture files to include top level of 1 only isn't enough.

can we get some unit tests for this trace processor?

I also think we should be sure we have some partial flushing tests that validate how top level is set (these tests _should_ already exist, it is just a matter of adding proper assertions to them and/or we can add some snapshot tests for partial flushing if we wanted)

File: tests/snapshots/tests.contrib.pyramid.test_pyramid.test_simple_pyramid_app_endpoint.json [hunk 7]
```diff
@@ -17,6 +17,7 @@
     "metrics": {
       "_dd.agent_psr": 1.0,
       "_dd.measured": 1,
+      "_dd.top_level": 1,
```
Comment (brettlangdon) [inline tests/snapshots/tests.contrib.pyramid.test_pyramid.test_simple_pyramid_app_endpoint.json:20]: is this a new snapshot file from upstream?

File: tests/snapshots/tests.contrib.rq.test_rq.test_enqueue_distributed_tracing_enabled_None_worker_service_custom-worker-service.json [hunk 8]
```diff
@@ -38,7 +38,7 @@
          "runtime-id": "79fea89591ef49e8b01debfbbd52f789"
        },
        "metrics": {
-         "_dd.top_level": 0,
+         "_dd.top_level": 1,
```
Comment (brettlangdon) [inline tests/snapshots/tests.contrib.rq.test_rq.test_enqueue_distributed_tracing_enabled_None_worker_service_custom-worker-service.json:41]: 👍🏻

File: tests/snapshots/tests.integration.test_integration_snapshots.test_tracer_trace_across_fork.json [hunk 9]
```diff
@@ -30,7 +30,7 @@
          "runtime-id": "409d5f5531a54659b762419af6a3abc5"
        },
        "metrics": {
-         "_dd.top_level": 0,
+         "_dd.top_level": 1,
```
Comment (brettlangdon) [inline tests/snapshots/tests.integration.test_integration_snapshots.test_tracer_trace_across_fork.json:33]: this is the trade off we are making where this will get marked as top level since the it is the local root in the fork.

the service naming being null is... potentially a problem. cc @Kyle-Verhoog

File: ddtrace/internal/processor/trace.py [hunk 10]
```diff
@@ -68,6 +68,33 @@ def process_trace(self, trace):
...
 
+@attr.s
+class TraceTopLevelSpanProcessor(TraceProcessor):
+    """Processor marks spans as top level"""
+
+    def process_trace(self, trace):
+        # type: (List[Span]) -> Optional[List[Span]]
+        """Mark a span in a trace as top level if:
+         1. Span has a different service name than it's parent
+         2. Span is a local root
```
Comment (brettlangdon) [inline ddtrace/internal/processor/trace.py:79]:
```suggestion
         1. Span is a local root
         2. Span has a different service name than it's parent
```

Just a small nit to swap the order to match the if statement below, e.g. 1. is more important than 2.

File: tests/tracer/test_processors.py [hunk 11]
```diff
@@ -216,3 +217,76 @@ def test_aggregator_partial_flush_2_spans():
...
+def test_trace_top_level_span_processor_orphan_span():
+    """Trace chuck does not contain parent span"""
+    trace_processors = TraceTopLevelSpanProcessor()
+    parent = Span(None, "span_not_in_trace", service="service")
+    parent._local_root = parent
+
+    orphan_span = Span(None, "orphan_span", service="service")
+    orphan_span._parent = parent
+    orphan_span.parent_id = orphan_span._parent.span_id
+    trace_processors.process_trace([Span(None, "span1"), orphan_span])
```
Comment (brettlangdon) [inline tests/tracer/test_processors.py:262]: is `Span(None, "span1")` needed here?

File: tests/tracer/test_processors.py [hunk 12]
```diff
@@ -216,3 +217,76 @@ def test_aggregator_partial_flush_2_spans():
...
+    assert child2.get_metric("_dd.top_level") == 1
+
+
+def test_trace_top_level_span_processor_orphan_span():
+    """Trace chuck does not contain parent span"""
+    trace_processors = TraceTopLevelSpanProcessor()
+    parent = Span(None, "span_not_in_trace", service="service")
+    parent._local_root = parent
+
+    orphan_span = Span(None, "orphan_span", service="service")
```
Comment (brettlangdon) [inline tests/tracer/test_processors.py:259]: for this, and all children spans, do we need to be setting `_local_root` explicitly on the spans?

File: tests/snapshots/tests.integration.test_integration_snapshots.test_tracer_trace_across_fork.json (hunk 9 above)
Comment (Kyle-Verhoog) [reply tests/snapshots/tests.integration.test_integration_snapshots.test_tracer_trace_across_fork.json:33]: service being null here is because we never set one in the test case, we probably should though to ensure that it does (or doesn't get propagated).

I don't think it does today which isn't great...

File: tests/tracer/test_processors.py (hunk 11 above)
Comment (mabdinur) [reply tests/tracer/test_processors.py:262]: Nope I just added it here to ensure that the trace processor does not process just the first span in a trace but iterates to the end. I could remove it if it makes this test confusing

File: tests/tracer/test_processors.py (hunk 12 above)
Comment (mabdinur) [reply tests/tracer/test_processors.py:259]: Good point. We could but it wouldn't affect the outcome of any of the tests. If `Span._local_root = None` do we assume that it is a local root? I can make this change if the current unit test is confusing

Comment (brettlangdon) [reply tests/tracer/test_processors.py:259]: I don't think we'd ever end up in a case (using `tracer.trace`) where `Span._local_root = None` right?

in which case how can we be sure we are properly testing realistic scenarios?

File: tests/tracer/test_processors.py (hunk 11 above)
Comment (brettlangdon) [reply tests/tracer/test_processors.py:262]: that should be a separate test case.

for example, if someone were to remove this later because it doesn't seem related to testing of the orphan span, then you would lose your validation that process_trace iterates through all of the spans.

we can try to keep each unit test case to fully test one specific behavior. if we have 10 scenarios we want to test, we should have 10 tests, try not to merge them together.

File: tests/tracer/test_processors.py [hunk 13]
```diff
@@ -216,3 +217,84 @@ def test_aggregator_partial_flush_2_spans():
...
+def test_trace_top_level_span_processor_no_parent_span_object():
+    """Parent span and child span have the different service names and Span._parent is None"""
+    trace_processors = TraceTopLevelSpanProcessor()
+    parent = Span(None, "span_not_in_trace", service="service")
+    parent._local_root = parent
+
+    span_wo_parent_span_obj = Span(None, "span_wo_parent_span_obj", service="new_service_name")
+    span_wo_parent_span_obj.parent_id = parent.span_id
+    span_wo_parent_span_obj._local_root = parent
+    span_wo_parent_span_obj._parent = None
```
Comment (brettlangdon) [inline tests/tracer/test_processors.py:282]: I don't think this is something that can actually happen in reality.

That is, if you have a hard reference to the parent span as `_local_root` but not `_parent`.

**Review by brettlangdon — CHANGES_REQUESTED** [review]
Test case missing is updated assertions for partial flushing to ensure `_dd.top_level` is set as we expect.

File: tests/snapshots/tests.contrib.pyramid.test_pyramid.test_simple_pyramid_app_endpoint.json (hunk 7 above)
Comment (mabdinur) [reply tests/snapshots/tests.contrib.pyramid.test_pyramid.test_simple_pyramid_app_endpoint.json:20]: Yes, I believe Zach generated this file while working on the pyramid escalation.

File: tests/tracer/test_processors.py (hunk 13 above)
Comment (mabdinur) [reply tests/tracer/test_processors.py]: Initially this test was designed to mock the flow of distributed traces and partial flushing. I'll remove this test and add a test case which uses `tracer.trace` and `tracer.configure(partial_flush_enabled=True)`

File: ddtrace/internal/processor/trace.py [hunk 14]
```diff
@@ -68,6 +68,38 @@ def process_trace(self, trace):
         return None
 
 
+@attr.s
+class TraceTopLevelSpanProcessor(TraceProcessor):
+    """Processor marks spans as top level"""
+
+    counter = 0
```
Comment (brettlangdon) [inline ddtrace/internal/processor/trace.py:75]: what is this used for?

File: tests/tracer/test_processors.py [hunk 15]
```diff
@@ -216,3 +218,79 @@ def test_aggregator_partial_flush_2_spans():
...
+def test_trace_top_level_span_processor_orphan_span():
+    """Trace chuck does not contain parent span"""
+
+    tracer = Tracer()
+    with tracer.trace("parent"):
+        with tracer.trace("orphan span") as orphan_span:
+            orphans_parent = Span(None, "span_wo_parent_span_obj")
+            orphan_span._parent = orphans_parent
+            orphan_span._local_root = orphans_parent
+            orphan_span.parent_id = orphans_parent.span_id
```
Comment (brettlangdon) [inline tests/tracer/test_processors.py:281]: you can probably get the same effect with:

```python
with tracer.trace("parent") as parent:
    pass

with tracer.start_span("orphan span", child_of=parent) as orphan_span:
    pass
```

File: tests/tracer/test_processors.py [hunk 16]
```diff
@@ -216,3 +218,79 @@ def test_aggregator_partial_flush_2_spans():
     assert writer.pop() == [child1, child2]
     parent.finish()
     assert writer.pop() == [parent]
+
+
+def test_trace_top_level_span_processor_partial_flushing():
+    """Parent span and child span have the same service name"""
+    tracer = Tracer()
```
Comment (brettlangdon) [inline tests/tracer/test_processors.py:225]: do we need to use a dummy writer here or something?

Comment (mabdinur) [reply tests/tracer/test_processors.py:225]: Yes, right now the tests pass but they log this warning `o handlers could be found for logger ddtrace.internal.writer`

Comment (brettlangdon) [reply tests/tracer/test_processors.py:225]: probably because it is failing to send the traces and it is trying to log the error.

if we can use a DummyWriter or something to ensure we don't actually try to send these traces it would be best. what if there is an agent running? will the performance/result be different? or we might just send unnecessary test traces to the backend.

File: tests/tracer/test_processors.py (hunk 13 above)
Comment (mabdinur) [reply tests/tracer/test_processors.py]: debugging I removed it

File: ddtrace/internal/processor/trace.py (hunk 14 above)
Comment (mabdinur) [reply ddtrace/internal/processor/trace.py]: Debugging. I removed it

**Review by brettlangdon — DISMISSED** [review]

File: ddtrace/internal/processor/trace.py [hunk 17]
```diff
@@ -68,6 +68,33 @@ def process_trace(self, trace):
...
+            AND parent span in not in the trace chunk
+        """
+
+        span_ids = {span.span_id for span in trace}
+        for span in trace:
+            if span is span._local_root:
+                span.set_metric("_dd.top_level", 1)
+            elif span._parent and span.service != span._parent.service:
+                span.set_metric("_dd.top_level", 1)
+            elif span.parent_id and span.parent_id not in span_ids:
```
Comment (P403n1x87) [inline ddtrace/internal/processor/trace.py:92]: Do we have an idea of the performance impact of this trace processor? Would it save us much if we memoised the result of `span.parent_id not in span_ids`?

Comment (brettlangdon) [reply ddtrace/internal/processor/trace.py:92]: This is an interesting question, I suppose it depends on whether it is a wide or deep trace.

I'd say wide traces (a lot of spans with the same parent id) are fairly rare. Usually it only happens when there is a function that makes a lot of database queries. But for that we are looking at an upper bound of about 20k child spans for the parent. Otherwise the average case is probably much closer to 1, I'd say definitely <10 on average.

Comment (P403n1x87) [reply ddtrace/internal/processor/trace.py:92]: I see. I think it's still worth being aware of what the cost is, perhaps compared to the other processors, just to have a rough idea. Also, forget about the memoisation, because I just happened to realise that `span_ids` is a `set`, so it's already memoised 😞 .

File: ddtrace/internal/processor/trace.py [hunk 18]
```diff
@@ -68,6 +68,33 @@ def process_trace(self, trace):
...
+
+    def process_trace(self, trace):
+        # type: (List[Span]) -> Optional[List[Span]]
+        """Mark a span in a trace as top level if:
+         1. Span is a local root
+         2. Span has a different service name than its parent
+
+        Explicitly set top level to zero/false if:
+         1. Span has a truthy parent_id (not zero or None)
+            AND parent span in not in the trace chunk
```
Comment (majorgreys) [inline ddtrace/internal/processor/trace.py:83]: This explanation could be clearer: high level what is a "top level span"? What is its functional use case? What are the values and how should the be interpreted? This docstring is closer to a specification, which can certainly be helpful here to keep.

Comment (mabdinur) [reply ddtrace/internal/processor/trace.py:83]: I will add the following description: A span is "top level" if it is the first span in a trace (root span), the first span wrapping a service (span.service != parent.service), or it is the top-most span created within a process (local root). The "top level" metric will be used by the agent to calculate trace metrics and determine how spans should be displaced in the UI. If this metric is not set by the tracer the first span in a trace chunk will be marked as top level.

**Review by brettlangdon — APPROVED** [review]

**Review by P403n1x87 — APPROVED** [review]

File: tests/tracer/test_processors.py [hunk 19]
```diff
@@ -216,3 +218,85 @@ def test_aggregator_partial_flush_2_spans():
     assert writer.pop() == [child1, child2]
     parent.finish()
     assert writer.pop() == [parent]
+
+
+def test_trace_top_level_span_processor_partial_flushing():
```
Comment (Kyle-Verhoog) [inline tests/tracer/test_processors.py:223]: These tests using a `Tracer` are probably better suited to be in `test_tracer.py` but I don't have a strong opinion

File: tests/tracer/test_processors.py [hunk 20]
```diff
@@ -216,3 +218,85 @@ def test_aggregator_partial_flush_2_spans():
...
+def test_trace_top_level_span_processor_orphan_span():
+    """Trace chuck does not contain parent span"""
+
+    tracer = Tracer()
+    tracer.configure(writer=DummyWriter())
+
+    with tracer.trace("parent") as parent:
+        pass
+
+    with tracer.start_span("orphan span", child_of=parent) as orphan_span:
```
Comment (Kyle-Verhoog) [inline tests/tracer/test_processors.py:286]: not really an orphan span if it has a parent 😆 but the test is good aside of the naming

**Review by Kyle-Verhoog — APPROVED** [review]
👏

File: tests/tracer/test_processors.py (hunk 20 above)
Comment (mabdinur) [reply tests/tracer/test_processors.py:286]: But the parent was terminated before the span was born. Now the span is all alone. Is the agent going to take care of it  😢

Comment (mabdinur) [reply tests/tracer/test_processors.py:286]: @Kyle-Verhoog

Comment (Kyle-Verhoog) [reply tests/tracer/test_processors.py:286]: the parent lives on in our hearts 🙌 🙌 ✊ ✊ 🙏 🙏
